units/migrate_volume_job.go

```diff
@@ -0,0 +1,286 @@
+func (j *volumeMigrationJob) stopInitialHost(ctx context.Context) {
+	// Stop host if this operation has not been attempted yet
+	if j.SpawnhostStopJobID == "" {
+		ts := utility.RoundPartOfMinute(1).Format(TSFormat)
+		stopJob := NewSpawnhostStopJob(j.initialHost, evergreen.User, ts)
+		err := j.env.RemoteQueue().Put(ctx, stopJob)
+		if err != nil {
+			j.AddRetryableError(err)
+			return
+		}
+		grip.Info(message.Fields{
+			"message":         "stopping initial host",
+			"job_id":          j.ID(),
+			"initial_host_id": j.InitialHostID,
+		})
+		j.SpawnhostStopJobID = stopJob.ID()
+		j.AddRetryableError(errors.Errorf("initial host '%s' not yet stopped", j.InitialHostID))
+		return
+	}
+
+	// Mark host as stopped
+	if j.initialHost.Status == evergreen.HostStopped {
+		grip.Info(message.Fields{
+			"message":         "initial host stopped",
+			"job_id":          j.ID(),
+			"initial_host_id": j.InitialHostID,
+		})
+		return
+	}
+
+	// If host is not stopped, verify whether the job is still in progress and error if it completed without stopping.
+	stopJob, foundJob := j.env.RemoteQueue().Get(ctx, j.SpawnhostStopJobID)
+	if !foundJob {
+		j.AddError(errors.Errorf("spawnhost-stop job '%s' could not be found", j.SpawnhostStopJobID))
+		return
+	}
+
+	if stopJob.Status().Completed {
+		j.AddError(errors.Errorf("host '%s' failed to stop", j.SpawnhostStopJobID))
+		return
+	}
+
+	j.AddRetryableError(errors.Errorf("initial host '%s' not yet stopped", j.InitialHostID))
+}
+
+// startNewHost attempts to start a new host with the volume attached.
+func (j *volumeMigrationJob) startNewHost(ctx context.Context) {
+	// Ensure volume has been detached
+	if j.volume.Host == j.InitialHostID {
+		j.AddRetryableError(errors.New("volume still attached to host"))
+		return
+	}
+
+	j.ModifyOptions.HomeVolumeID = j.VolumeID
+	intentHost, err := cloud.CreateSpawnHost(ctx, j.ModifyOptions, j.env.Settings())
+	if err != nil {
+		j.AddRetryableError(errors.Wrapf(err, "creating new intent host"))
+		return
+	}
+
+	if err := intentHost.Insert(); err != nil {
+		j.AddError(errors.Wrap(err, "inserting new intent host"))
+		return
+	}
+	grip.Info(message.Fields{
+		"message":        "new intent host created",
+		"job_id":         j.ID(),
+		"intent_host_id": intentHost.Id,
+	})
+	j.NewHostID = intentHost.Id
+}
+
+// finishJob marks the job as completed and attempts some additional cleanup if this is the job's final attempt.
+func (j *volumeMigrationJob) finishJob(ctx context.Context) {
+	// Amboy's ShouldRetry() method does not account for reaching MaxAttempts, so manually determine if the current attempt is the last.
+	retryInfo := j.RetryInfo()
+	isFinalAttempt := retryInfo.CurrentAttempt >= (retryInfo.MaxAttempts - 1)
+	if !retryInfo.ShouldRetry() || isFinalAttempt {
+		// If a new host was never created, restart initial host with volume attached.
+		if j.NewHostID == "" {
+			if hasErr := j.restartInitialHost(ctx); hasErr {
```

Comment:
Do we want to restart the initial host in case the migration fails? I imagine that the user might just want to retry the migration, so re-stopping the just-restarted host might be redundant.

Reply:
I had this thought too, but some spec reviewers thought it'd be more useful to leave users with a usable host in case of failure, so I went with this.

Reply:
I agree that a usable host would be nice, but I think it might be hard to resuscitate the old host at this point. Even if it can restart, by the time it's reached this step, it's already detached the home volume which probably has most of the host's files, so the user would have to manually re-attach the volume as well.

Reply:
This job does reattach the home volume in this step though right?

Reply:
Hmm yeah, looking at it, I guess it does try to attach the volume. I'm trying to avoid this sort of expensive end-of-job fixup operations because it's not a reliable mechanism to fix the old host.

Reply:
Would there be a way for users to reattach the home volume manually in error @sophstad ? Or is it something we'd consider adding a button for?

Reply:
> I'm trying to avoid this sort of expensive end-of-job fixup operation though because it's not a reliable mechanism to fix the old host and it's probably happening in a case that can't be fixed

This makes a lot of sense! And @ablack12 if we don't go ahead with the restarting step, we could allow users to remount their volume with the normal UI mount button. The initial host would be unusable without the volume attached at this point (see [EVG-7879](https://jira.mongodb.org/browse/EVG-7879)), so it would maybe make sense to just terminate it?

Reply:
I'd rather we not terminate it, if they are still able to remount their volume. Maybe we could set the expiration date back to expirable though so it doesn't stick around if they don't want it, similar to if the job does actually work? Also I wonder if we could add a UI warning label thing like "no home volume attached" (that might be too far though; it would also be sufficient to just have trying to start it again return an error)

Reply:
People do get really testy about us deleting their hosts when they don't want them deleted so I think we'd better err on the side of caution with it.

Reply:
Annie is out now, but to sum up the takeaway here, I'm going to:

- Remove the recovery portion of the job that attempts to restart the initial host in case of failure. The job will finish with a stopped initial host and unmounted volume that can then be attached and restarted. I don't think we need to touch the expiry date as it seems inconsequential/maybe more confusing.
- Remove the final termination-upon-success step, and instead set the initial host to expire in 24 hours (as discussed in [this comment](https://github.com/evergreen-ci/evergreen/pull/5966#discussion_r982824998)

@Kimchelly let me know if you have any objections!

Reply:
This sounds good to me! I have two questions about the expiration:
* If it's set to 24 hours, does the user get a notification that their host will be expired soon? It would be nice to make sure that the user knows it's going to go away soon.
* Is it worth logging an event for the host to tell the user that it failed to migrate and will be expired soon?